src/analysis_and_optimization/Mem_pattern.ml

```diff
@@ -0,0 +1,602 @@
+          Set.Poly.union_list (List.map ~f:(query_expr_loop in_loop) exprs)
+        else ret_demotable_types )
+  | CompilerInternal (_ : Internal_fun.t) -> Set.Poly.empty
+  | UserDefined ((_ : string), (_ : bool Fun_kind.suffix)) ->
+      ret_demotable_types
+
+(**
+ * Find indices on Matrix and Vector types that perform single 
+ *  cell access. Returns true if it finds
+ * a vector, row vector, matrix, or matrix with single cell access
+ * as well as an array of any of the above that is accessing the 
+ * inner matrix types cell.
+ *)
+let rec is_uni_eigen_loop_indexing (ut : UnsizedType.t)
+    (index : 'a Expr.Fixed.t Index.t list) =
+  match (ut, index) with
+  | (UnsizedType.UArray t | UFun (_, ReturnType t, _, _)), index -> (
+    match List.tl index with
+    | Some cut_list -> is_uni_eigen_loop_indexing t cut_list
+    | None -> is_uni_eigen_loop_indexing t [Index.All] )
+  | (UVector | URowVector), [Index.Single (_ : 'a Expr.Fixed.t)] -> true
+  | UMatrix, [Single (_ : 'a Expr.Fixed.t); Single (_ : 'a Expr.Fixed.t)] ->
+      true
+  (* None of the below contain single cell access*)
+  | (UReal | UInt | UMathLibraryFunction | UFun (_, Void, _, _)), _ -> false
+  | (UVector | URowVector), _ -> false
+  | UMatrix, _ -> false
+
+(**
+ * Specialization to find demotable (SoA -> AoS) objects
+ *  that carries logic on whether we are in a for loop or not.
+ * (TODO: Could be combined with other query demotable above.)
+ *)
+let rec query_demotable_exprs_loop in_loop Expr.Fixed.({pattern; _}) =
+  let query_expr = query_demotable_exprs_loop in_loop in
+  match pattern with
+  | FunApp (kind, (exprs : Expr.Typed.Meta.t Expr.Fixed.t list)) ->
+      query_demotable_funkinds_loop query_demotable_exprs_loop in_loop kind
+        exprs
+  | Var (_ : string) | Lit ((_ : Expr.Fixed.Pattern.litType), (_ : string)) ->
+      Set.Poly.empty
+  | TernaryIf (predicate, texpr, fexpr) ->
+      Set.Poly.union
+        (Set.Poly.union (query_expr predicate) (query_expr texpr))
+        (query_expr fexpr)
+  | Indexed (Expr.Fixed.({pattern= Var name; meta= {type_; _}}), indexed) ->
+      let query_index = map_index query_expr in
+      if in_loop && is_uni_eigen_loop_indexing type_ indexed then
+        Set.Poly.union (Set.Poly.singleton name)
+          (Set.Poly.union_list (List.map ~f:query_index indexed))
+      else Set.Poly.union_list (List.map ~f:query_index indexed)
+  | Indexed (expr, indexed) ->
+      let query_index = map_index query_expr in
+      Set.Poly.union (query_expr expr)
+        (Set.Poly.union_list (List.map ~f:query_index indexed))
+  | EAnd (lhs, rhs) | EOr (lhs, rhs) ->
+      Set.Poly.union (query_expr lhs) (query_expr rhs)
+
+(**
+ * Specialization of the above to handle demotion while carrying 
+ * information on whether we have entered a for loop.
+ * (TODO: We can probably coerce this and the 
+ *  other query demotable function together)
+ * 
+ *)
+let rec query_demotable_stmts_loop in_loop Stmt.Fixed.({pattern; _}) =
```

Comment:
It seems like you have two copies of some big functions, one that's loop-aware and one that isn't. You're also double-checking for AoS functions since you do it in `query_demotable_stmts_loop` for the initial variables and then again during the monotone analysis in `query_demotable_stmt` inside `gen_aos_variables`.

I think we could avoid those issues if we check for the loop pattern separately from the AoS functions, and check the AoS functions either for initial variables or in the gen function.

Reply:
My suggestion would be something that uses `fold_stmts` like this:

```
(* Collect sets of things from expressions and statements inside of a loop, using the loop variable *)
let collect_in_loops ~collect_looped_expr ~collect_looped_stmt ~(init : 'c)
    (stmts : Stmt.Located.t List.t) : 'c =
  fold_stmts ~init
    ~take_expr:(fun c _ -> c)
    ~take_stmt:(fun c stmt ->
      match stmt.pattern with
      | Stmt.Fixed.Pattern.For {loopvar; body; _} ->
          fold_stmts ~init:c
            ~take_expr:(fun c expr ->
              Set.Poly.union c (collect_looped_expr loopvar expr) )
            ~take_stmt:(fun c stmt ->
              Set.Poly.union c (collect_looped_stmt loopvar stmt) )
            [body]
      | _ -> c )
    stmts

(* Collect names of eigen variables that are single indexed inside a loop *)
let collect_single_indexed_in_loops stmts =
  let collect_single_indexed_expr loopvar Expr.Fixed.({pattern; meta}) =
    match pattern with
    | Expr.Fixed.Pattern.Indexed (Expr.Fixed.({pattern = Var var; _}), indices)
      when is_uni_eigen_loop_indexing meta.type_ indices ->
        Set.Poly.singleton var
    | _ -> Set.Poly.empty
  in
  let collect_single_indexed_lhs loopvar stmt =
    match stmt.pattern with
    | Assignment ((var, type_, lhs), _)
      when is_uni_eigen_loop_indexing expr.meta.type_ indices ->
        Set.Poly.singleton var
    | _ -> Set.Poly.empty
  in
  collect_in_loops ~init:Set.Poly.empty
    ~collect_looped_expr:collect_single_indexed_expr
    ~collect_looped_stmt:collect_single_indexed_expr stmts
```

Reply:
You could also do the above inside the monotone framework by converting to the typical MIR format with `Mir_utils.stmt_loc_of_stmt_loc_num`, but I see initial variables work too

Reply:
> It seems like you have two copies of some big functions, one that's loop-aware and one that isn't. You're also double-checking for AoS functions since you do it in query_demotable_stmts_loop for the initial variables and then again during the monotone analysis in query_demotable_stmt inside gen_aos_variables.

So one big thing I realized is that I actually don't need to look at expressions at all during the monotone framework phase! Since we are only querying expressions for finding objects that

1.  Are used in functions that don't support SoA
2.  Use single cell indexing or assignment

we can actually do both of those while collecting the initial variables. So I think that means I can delete a bunch of code 

I'm fine with breaking out the for loop check and functions check. It requires two passes but should be easier to read. If we need to start caring about performance it should be pretty easy to fold those back into one big pass

For the fold_stmts version you posted above, does that work for searching inner stmts? Like if I have a `Block` with a `For` inside of it, can those fold statements handle that?

Reply:
> You could also do the above inside the monotone framework by converting to the typical MIR format with Mir_utils.stmt_loc_of_stmt_loc_num, but I see initial variables work too

I think I'd like to limit the use of the monotone framework to only the parts that def need it like the assignment statements. For the rules above I can just iterate over the mir without needing anything fancy, no need to build a birdhouse with a nailgun, simplest tool for the job, etc

Reply:
> So one big thing I realized is that I actually don't need to look at expressions at all during the monotone framework phase

Except, I guess,
- Inferring the memory pattern of the RHS of an assignment
- Getting the set of variables that are implied to be AoS when the LHS is AoS

> For the fold_stmts version you posted above, does that work for searching inner stmts? Like if I have a Block with a For inside of it, can those fold statements handle that?

Yep!

> For the rules above I can just iterate over the mir without needing anything fancy, no need to build a birdhouse with a nailgun, simplest tool for the job, etc

Yeah, but you could also argue that since you're already using the monotone framework, so it's a matter of just monotone framework vs. traversal then monotone framework. But I see what you mean and it's reasonable either way